atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -199,21 +202,56 @@ private SweepResults doRun(TableReference tableRef,
     }
 
     /**
-     * Returns the number of deleted (cell, timestamp) pairs.
+     * Partitions batches of cells down to 2 * deleteBatchSize blocks per sweep deletion call.
+     * Returns the number of blocks that were deleted.
      */
-    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType) {
-        Multimap<Cell, Long> startTimestampsToSweepPerCell = ArrayListMultimap.create();
-        List<Cell> sentinels = Lists.newArrayList();
+    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType,
+            int deleteBatchSize) {
+        int numberOfSweptCells = 0;
+
+        // Previously we've grouped cells to have at least deleteBatchSize blocks per batch. Therefore we expect most
+        // of the batches to be slightly bigger than deleteBatchSize. Partitioning such batches with deleteBatchSize as
+        // a limit make us have a small second batch. Therefore, we set our limit per batch to 2 * deleteBatchSize.
+        deleteBatchSize = 2 * deleteBatchSize;
+
+        Multimap<Cell, Long> currentBatch = ArrayListMultimap.create();
+        List<Cell> currentBatchSentinels = Lists.newArrayList();
+
         for (CellToSweep cell : batch) {
-            startTimestampsToSweepPerCell.putAll(cell.cell(), TDecorators.wrap(cell.sortedTimestamps()));
             if (cell.needsSentinel()) {
-                sentinels.add(cell.cell());
+                currentBatchSentinels.add(cell.cell());
+            }
+
+            Long[] sortedTimestampsArray = ArrayUtils.toObject(cell.sortedTimestamps().toArray());
```

Comment:
To avoid extra allocations, you could use Longs.asList() from Guava. It creates a view of the array and has a O(1) subList impl.

Or, to save another extra copy (sortedTimestamps().toArray()), you can simply create the collection decorator using TDecorators.wrap(). The resulting collection seems to have a regular O(1) subList impl from the AbstractList.